File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
n, m = map(int, input().split())

arr = list(map(int, input().split()))

start = 0
end = max(arr)

while start <= end:
    trees = 0
    mid = (start + end) // 2
    for elem in arr:
        if elem > mid:
            trees += elem - mid
    if trees >= m:
        start = mid + 1
    else:
        end = mid - 1

print(end)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.swift
let nm = readLine()!.split(separator: " ").map { Int($0)! }
let n = nm[0], m = nm[1]

let arr = readLine()!.split(separator: " ").map { Int($0)! }

var start = 0, end = arr.max()!

binarySearch()

func binarySearch() {
    while start <= end {
        let mid = (start + end) / 2
        let sum = arr.reduce(0) { $0 + ($1 > mid ? $1 - mid : 0) }
        if sum >= m {
            start = mid + 1
        } else {
            end = mid - 1
        }
    }
    print(end)
}
